python/ml4ir/applications/ranking/data/scripts/create_dataset.py

```diff
@@ -0,0 +1,227 @@
+    """
+    1. Loads example data
+    2. Builds specified synthetic data size by sampling from example data
+    3. Adds catastrophic failures specifically
+    4. For now, write out to CSV. In future could return df directly
+    """
+
+    # Setup logging
+    logs_dir: str = os.path.join('logs', 'test_run_id')
+    file_io.make_directory(logs_dir, clear_dir=True, log=None)
+    logger: Logger = setup_logging()
+    logger.info('Logging initialized. Saving logs to : {}'.format(logs_dir))
+    logger.info('Run ID: {}'.format('test_run_id'))
+
+    try:
+        # Set seeds
+        set_seeds(random_state)
+        logger.info('Set seeds with initial random state {}'.format(random_state))
+
+        # Load and parse feature config
+        feature_config: FeatureConfig = parse_config(tfrecord_type='', feature_config=feature_config, logger=logger)
+        logger.info("Feature config parsed and loaded")
+
+        # Create output location
+        file_io.make_directory(out_dir, log=logger)
+        out_file = os.path.join(out_dir, 'synthetic_data_{}.csv'.format(dt.datetime.now().strftime('%Y%m%d-%H%M%S')))
+
+        # Build data
+        seed_data = load_seed_data(data_dir, logger)
+
+        df_synthetic = fill_data(logger,
+                                 seed_data,
+                                 max_num_records,
+                                 feature_config,
+                                 feature_highval,
+                                 feature_num_results,
+                                 num_samples)
+        file_io.write_df(df_synthetic, outfile=out_file, index=False)
+        logger.info('Synthetic data created! Location: {}'.format(out_file))
+        return df_synthetic
+
+    except Exception as e:
+        logger.error("!!! Error creating synthetic data: !!!\n{}".format(str(e)))
+        traceback.print_exc()
+        return
+
+
+def setup_logging():
+    return logging_utils.setup_logging(reset=True, file_name='test.txt', log_to_file=True)
+
+
+def set_seeds(random_state):
+    # for repeatability
+    np.random.seed(random_state)
+    random.seed(random_state)
+
+
+def load_seed_data(data_dir, logger):
+    # Start with loading in CSV format only
+    if file_io.path_exists(data_dir):
+        dfs = file_io.get_files_in_directory(data_dir)
+        return file_io.read_df_list(dfs)
+    else:
+        logger.error("Error! Data directory must exist and be specified")
+
+
+def generate_key(seed_dict, query_key, log=None):
+    seed_id = str(random.sample(seed_dict[query_key], 1)[0])
+    q_id = ''.join(random.sample(seed_id, len(seed_id)))
+    if q_id in set(seed_dict.keys()):
+        if log:
+            log.info('Generated query key is duplicate, regenerating. If happens repeatedly, query key may be too short')
+        return generate_key(seed_dict, query_key)
+    return q_id
+
+
+def filter_nres(seed_data, feature_query_key, feature_rank, feature_nres, max_num_records, logger):
+    nres = 'num_results_calc'
```

Comment:
@calliebradley [Minor] Probably better to move all hardcoded string constants to a separate enum/class at the top. Something like here -> https://github.com/salesforce/ml4ir/blob/master/python/ml4ir/base/model/relevance_model.py#L24

Reply:
Jake requested it not be a class calling it overly structured ;) I can move these to the top / all-caps.

Reply:
I get that. Not suggesting that we make the synthetic data creation a class. Just an enum for the constants. There is no enum data structure in python. So you can use a class. Or you can even use https://docs.python.org/3/library/enum.html